ENT-3831: Added new thread safe queue data structure

karlhto opened the pull request:
Can push, pop, block before popping if queue is empty, return
amount of elements, capacity, copy, etc. Can also be created
as a non-thread safe queue, if that is needed.

Ticket: ENT-3831
Changelog: None

olehermanse commented:
@cf-bottom Please test this in jenkins :)

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1111)](https://ci.cfengine.com//job/pr-pipeline/1111/)

https://ci.cfengine.com/job/pr-pipeline/1111/

olehermanse commented:
IMO, it is better to introduce this as a new data structure `ThreadedQueue` first, then phase out the old one. Also, have you considered way to use the data structure without mutex? Like if I want to use queue in a single thread, and am worried about performance?

karlhto commented:
Yeah, you might be right. As for a non-mutex version, how would that best be done? Is there maybe a way to do it without adding if-checks before each locking? Probably not

karlhto commented:
@cf-bottom jenkins? exotics!

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1112)](https://ci.cfengine.com//job/pr-pipeline/1112/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1112/

olehermanse commented:
@karlhto You could add `if`s or separate functions.

karlhto commented on libutils/queue.c:
```diff
@@ -23,121 +23,386 @@
+    Queue *new_queue = xmemdup(queue, sizeof(Queue));
+    new_queue->data = xmalloc(sizeof(void *) * queue->capacity);
+    memcpy(new_queue->data, queue->data,
+           sizeof(void *) * new_queue->capacity);
```
There was an issue with this that I discovered when creating StackCopy. Reallocating a memdup'ed pointer will give a runtime error. At least in Valgrind

karlhto commented:
@cf-bottom jenkins with exotics, please.

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1117)](https://ci.cfengine.com//job/pr-pipeline/1117/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1117/

karlhto commented:
Honestly a bit unsure about the ThreadLock wrapper functions, but I think it's nicer. What is not nicer is the locking not being uniform, like having to do `if (queue->lock)` instead of the functions when doing more than just locking.

olehermanse commented:
@karlhto This fails in tests, here is the relevant output:

```
02:24:34 ==================================================
02:24:34 Starting test: queue_test.c
02:24:34 ==================================================
02:24:34 test_basics: Starting test
02:24:34 test_basics: Test completed successfully.
02:24:34 test_destroy: Starting test
02:24:34 test_destroy: Test completed successfully.
02:24:34 All 2 tests passed
02:24:34 PASS: queue_test
02:24:34 Locking failure at threaded_queue.c:211 function ThreadedQueuePush! (pthread_mutex_lock: Resource deadlock avoided)==================================================
02:24:34 Starting test: threaded_queue_test.c
02:24:34 ==================================================
02:24:34 test_push_pop: Starting test
02:24:34 FAIL: threaded_queue_test
02:24:34 ==================================================
```

karlhto commented:
@olehermanse I already fixed that, just didn't trigger a new build.

@cf-bottom jenkins + exotics

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1119)](https://ci.cfengine.com//job/pr-pipeline/1119/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1119/

vpodzime commented on libutils/threaded_queue.c:
```diff
@@ -0,0 +1,461 @@
+    return item;
+}
+
+void *ThreadedQueuePopBlock(ThreadedQueue *queue)
```
It would also be great to have `ThreadedQueuePopN(queue, n, &n_popped)`, which would pop at most `n` elements at once (within a single lock-unlock "session"). If the items are easy to process, the locking overhead can be quite big if just one item is popped at a time.

karlhto replied:
>>> I think these two functions should be merged into a single one with the `timeout` argument determining what should happen -- return immediately (`0`), wait for `timeout` seconds or wait indefinitely (`-1`). And such a function needs to return `true/false` to indicate if it got a value or timed out.

Won't this inhibit how obvious it is to use? I would rather have two functions, where one returns immediately regardless, and the other takes in enums like `THREAD_BLOCK_INDEFINITELY` (-1), `THREAD_BLOCK_NONE` (0), and some other default value as a parameter.

Maybe I'm focusing on the wrong things?

vpodzime replied:
I think we are all capable of reading the function's documentation if it's there. So I don't think it's a big deal if we only have one function. But if you want to add really straightforward functions like `PopWait()` and `TryPop()`, I think they should be based on a generic one that has this **properly documented** versatile `int timeout` parameter.

vpodzime commented on tests/unit/threaded_queue_test.c:
```diff
@@ -0,0 +1,171 @@
+    assert_int_equal(ThreadedQueueCount(queue), 11);
+    assert_int_equal(ThreadedQueueCapacity(queue), 16);
+
+    ThreadedQueueDestroy(queue);
```
Would be nice to add comments to this test that would describe how the internal array looks like (or *is supposed to look like*) at those particular stages. IOW, explanations of what the particular blocks are testing.

vpodzime replied:
I still think these comments would be useful.

karlhto commented on libutils/threaded_queue.c:
```diff
@@ -0,0 +1,461 @@
+static inline void *_PopNonLock(ThreadedQueue *queue);
+static inline void _PushNonLock(ThreadedQueue *queue, void *item);
+static inline void _ThreadedQueueLock(pthread_mutex_t *lock);
+static inline void _ThreadedQueueUnlock(pthread_mutex_t *lock);
```
If you don't use a prefix like that, it isn't implicit that the function is only usable inside the scope of the file if you see it used inside another function. I think it's a nice convention, but I'm unsure of how much it is used in the rest of the codebase.

karlhto commented on libutils/threaded_queue.c:
```diff
@@ -0,0 +1,461 @@
+    _ThreadedQueueUnlock(queue->lock);
+
+    return empty;
+}
```
What would it do? Wait until it is empty, or wait until it is not empty?

karlhto commented:
Not entirely sure about what you mean with the review comment. The old queue is probably a lot slower due to cache misses and needing to allocate memory each time it adds a new element, since it is a linked list. Therefore it _does_ make sense to use a single-threaded version of this, but maybe what you mean is redoing the original queue struct to use arrays instead? Then you would remove the overhead of having to check if the lock exists each time you are trying to access the queue.

olehermanse commented:
> If you have a single thread that has a list/set/sequence of things to process, why would you not just use the Seq structure that we already have? I don't think that some artificial example of where a thread-unsafe FIFO structure would be needed is worth complicating the code.

I completely agree, but prefer a gradual transition. I.e. keep the old crappy unsafe queue, add this in a new file with a new name, and gradually revisit the places where the old queue is used, replacing it with the new queue and possibly making other fixes while reviewing that code.

vpodzime commented:
> replacing it with the new queue and possibly making other fixes while reviewing that code

Sure, but do we have a place where the old crappy unsafe queue is used in a single thread and without locks? If so, should such a place use a queue/FIFO?

olehermanse commented:
> Sure, but do we have a place where the old crappy unsafe queue is used in a single thread and without locks? If so, should such a place use a queue/FIFO?

@vpodzime

A) I don't know without searching, which is sort of my point, it's safer to do this in steps, introduce new data structure (unused) and then start using and replacing.

B) After searching, the old Queue is used in 1 place, in nova. So either a simultaneous PR is needed, or this new queue must have a different name.

@karlhto Open a PR for nova (report-collector.c) for this to be mergeable, (or change the name of this new Queue).

karlhto commented:
@olehermanse This is already mergeable. I changed it from Queue to ThreadedQueue a while ago.

I'll push fixes tomorrow, but I need to know what you and @vpodzime decide upon. I agree with the doing it in steps part, and we can remove unused functions later.

vpodzime commented on libutils/mutex.c:
```diff
@@ -69,3 +68,53 @@ void __ThreadUnlock(pthread_mutex_t *mutex,
+    }
+    else if (timeout == THREAD_BLOCK_NONE)
+    {
+        return 0;
```
This should still call `pthread_cond_timedwait()` and return a value saying if the condition was signalled or not.

karlhto replied:
I'm not sure if I agree. That will add unnecessary calls if you want to run the function without waiting for a signal.

Also, I think the signal is only sent to threads that are waiting, which means that it will never be in a situation where it will return that it has been signalled. Correct me if I'm wrong!

karlhto replied:
...I guess that option then becomes completely useless, just like not running the function at all. I think I will remove it and only use the `THREAD_BLOCK_INDEFINITELY`.

vpodzime commented on libutils/mutex.h:
```diff
@@ -27,12 +27,21 @@
 
+enum {
+    THREAD_BLOCK_INDEFINITELY = -1,
+    THREAD_BLOCK_NONE
```
I think these two should be macros.

vpodzime commented on libutils/mutex.c:
```diff
@@ -69,3 +68,53 @@ void __ThreadUnlock(pthread_mutex_t *mutex,
     }
 }
+
+int __ThreadWait(pthread_cond_t *pcond, pthread_mutex_t *mutex, int timeout,
```
Please add a documentation comment to this function and refer to the two important values of `timeout`.

vpodzime commented on libutils/threaded_queue.c:
```diff
@@ -0,0 +1,393 @@
+static void _ExpandIfNecessary(ThreadedQueue *queue);
+
+static inline void *_PopNonLock(ThreadedQueue *queue);
+static inline void _PushNonLock(ThreadedQueue *queue, void *item);
```
I think we don't need these two functions (and a "mode" of the structure that doesn't care about the locks). And they should not start with an underscore. They are `static`, that says it all.

karlhto commented on libutils/mutex.c:
```diff
@@ -69,3 +68,53 @@ void __ThreadUnlock(pthread_mutex_t *mutex,
     }
 }
+
+int __ThreadWait(pthread_cond_t *pcond, pthread_mutex_t *mutex, int timeout,
```
If you look at the other functions in mutex.c and mutex.h, they are like that as well due to having a macro that fills in `__line__`, `__func__`, etc.

olehermanse replied:
@karlhto Okay, but you are using it for other functions as well which are just static.

olehermanse replied:
`_DestroyRange`, `_ExpandIfNecessary`,  `_PopNonLock` etc.

karlhto commented on libutils/threaded_queue.c:
```diff
@@ -0,0 +1,393 @@
+        queue->ItemDestroy(queue->data[start]);
+
+        // in case start == end, start at second element in range
+        for (size_t i = start + 1; i != end; i++)
```
Start is destroyed outside the loop. I do it this way since `queue->tail` can be the same as `queue->head`. I think it makes more sense that end is excluded, since it then makes sense to pass in just `queue->head` and `queue->tail`.

karlhto commented on libutils/threaded_queue.c:
```diff
@@ -0,0 +1,393 @@
+    assert(queue != NULL);
+    if (start > queue->capacity || end > queue->capacity)
+    {
+        return;
```
This function already makes lots of assumptions about the data inside the internal array. Maybe this should be inside an `assert()` call, since it should _never_ be called with start and end indexes outside the capacity?

olehermanse replied:
@vpodzime likes ADA, so he might prefer that you do bounds checking at runtime ;)

karlhto commented:
@olehermanse @vpodzime What about just having one function for pushing that returns the current size, instead of both `ThreadedQueuePush` and `ThreadedQueuePushReportCount`? I feel like that is logical behaviour for a push function.

olehermanse commented:
@karlhto I think that's fine, but it sounds like it would break the compatibility with the other Queue implementation?

karlhto commented:
How? You don't have to use the return value for anything.
`QueueEnqueue(call_queue, contact);` -> `ThreadedQueuePush(call_queue, contact);`

karlhto commented:
@cf-bottom jenkins and exotics, please! 💯

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1305)](https://ci.cfengine.com//job/pr-pipeline/1305/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1305/

olehermanse commented:
> How? You don't have to use the return value for anything.
> `QueueEnqueue(call_queue, contact);` -> `ThreadedQueuePush(call_queue, contact);`

Makes sense, I thought the existing function had a boolean return value or something similar. Note that I said, "sounds like", because I didn't look up the source code before writing ;)

karlhto commented:
@cf-bottom jenkins with exotics

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1318)](https://ci.cfengine.com//job/pr-pipeline/1318/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1318/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1335)](https://ci.cfengine.com//job/pr-pipeline/1335/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1335/

karlhto commented:
@olehermanse @vpodzime Addressed most of your concerns. The man page suggested I checked for spurious wakeups with `pthread_cond_wait`, which means I have to reevaluate the predicate whenever it wakes up. I am unsure if the way I did it is safe, since the variables the predicates check for may be changed by another thread that is waiting for lock acquisition and acquires it before the "waiting" thread.

```
do {
    res = ThreadWait(queue->cond_empty, queue->lock, timeout);
    if (res != 0)
    {
        return false;
    }
} while (queue->size != 0);
// Reevaluate predicate to protect against spurious wakeups
```

Also, tests are needed for thread-specific stuff before this is production ready.

karlhto commented:
...also, HP-UX does not have `CLOCK_MONOTONIC`, which might be a problem if I want to accurately time something, since `CLOCK_REALTIME` is not as accurate.
As a temporary solution I defined `CLOCK_MONOTONIC` as `CLOCK_REALTIME` if the OS is HP-UX.

karlhto commented:
@olehermanse @vpodzime I have now added tests for threads, although it's possible they should be grouped in the stress testing directory instead.

The first thread test spawns threads that block until there are elements in the queue, where the elements are added by spawned threads that push to the queue. The second test fills the queue, spawns a thread that waits until the queue is empty, then empties the queue. The thread that was waiting then adds an element to the queue, which is read by the main thread.

karlhto commented:
@cf-bottom Jenkins! Exotics! Woo

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1380)](https://ci.cfengine.com//job/pr-pipeline/1380/)

https://ci.cfengine.com/job/pr-pipeline/1380/

karlhto commented:
I guess we're trying again
@cf-bottom Jenkins with exotics, please!

@olehermanse could you remove the previous attempt?

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1386)](https://ci.cfengine.com//job/pr-pipeline/1386/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1386/

karlhto commented:
@cf-bottom jenkins with exotics, please!

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1412)](https://ci.cfengine.com//job/pr-pipeline/1412/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1412/

karlhto commented on libutils/threaded_queue.c:
```diff
@@ -0,0 +1,504 @@
+                *data_array = NULL;
+                return 0;
+            }
+        } while (queue->size == 0);
```
I think it's possible that predicate rechecking can result in race conditions, since another asynchronous operation might empty the queue before this thread acquires the lock, which will result in the thread just continuing to wait.

karlhto replied:
> But the lock is acquired as part of the ThreadWait() call, right? So I don't think this point can ever be reached while not holding the lock.

I meant that: even though the thread has been signalled, the lock might be acquired by another thread before the signalled thread acquires it, which means that the predicate might be changed before this thread ever gets to check it, resulting in it just looping and waiting again. I suppose that's fine for ThreadedQueuePop, but it is probably _not_ fine for ThreadedQueueWaitEmpty, since you might have several threads that want to populate the queue when it has become empty.
I don't think there's a better way to protect against random wake-ups, so it will have to do.

vpodzime replied:
> I meant that: even though the thread has been signalled, the lock might be acquired by another thread before the signalled thread acquires it

The thread *being signalled* and the thread *acquiring the lock* happens at the same time, AFAICT.

karlhto commented:
@olehermanse @vpodzime Please review again now. Pretty sure it should run fine on all the test platforms.

karlhto commented:
@cf-bottom Jenkins with exotics

vpodzime commented:
@cf-bottom Jenkins with exotics

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1433)](https://ci.cfengine.com//job/pr-pipeline/1433/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1433/

vpodzime commented on libutils/mutex.c:
```diff
@@ -70,3 +70,49 @@ void __ThreadUnlock(pthread_mutex_t *mutex,
+            fflush(stdout);
+            fflush(stderr);
+            fsync(STDOUT_FILENO);
+            fsync(STDERR_FILENO);
```
@craigcomstock you have seen some issues with things like this, haven't you?

craigcomstock replied:
Oh yes. I had introduced fflush() and fsync() previously and it brought some builds/tests to a halt with apparent deadlocks maybe between processes or threads. I can't find the exact code but I think it was in logging.c LogToConsole(). I think that calling fflush() is sufficient where fsync() can cause deadlocks.

karlhto commented:
@cf-bottom Jenkins with exotics, please

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1466)](https://ci.cfengine.com//job/pr-pipeline/1466/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1466/

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1468)](https://ci.cfengine.com//job/pr-pipeline/1468/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1468/

karlhto commented:
@olehermanse Could you cancel the last build of the two above? :s

karlhto commented:
@cf-bottom Jenkins with exotics, please

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1475)](https://ci.cfengine.com//job/pr-pipeline/1475/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1475/

karlhto commented:
Maybe it will actually build on all instances now:
[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=1519)](https://ci.cfengine.com/job/pr-pipeline/1519/)

https://ci.cfengine.com/job/pr-pipeline/1519/

karlhto commented:
@olehermanse Any idea why all my builds end up being aborted?

olehermanse commented:
@karlhto It looks like the `queue_test` is hanging on RHEL 7. @vpodzime can confirm / elaborate.

vpodzime commented:
> @karlhto It looks like the `queue_test` is hanging on RHEL 7. @vpodzime can confirm / elaborate.

Yes, somehow. It actually looks like the test finishes and then it all hangs, but I didn't inspect it too much.

karlhto commented:
`threaded_queue_test` is executed right after, so that's probably the cause, then.

karlhto commented:
@cf-bottom Jenkins with exotics

Removed an unnecessary macro that will most likely not change anything. (:

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1530)](https://ci.cfengine.com//job/pr-pipeline/1530/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1530/

karlhto commented:
@cf-bottom Jenkins with exotics, please.

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1573)](https://ci.cfengine.com//job/pr-pipeline/1573/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1573/

olehermanse commented:
@cf-bottom I want to see another green build in jenkins, with exotics ;)

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1585)](https://ci.cfengine.com//job/pr-pipeline/1585/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1585/

vpodzime commented on libutils/threaded_queue.h:
```diff
@@ -0,0 +1,132 @@
+        If THREAD_WAIT_INDEFINITELY is specified, waits forever until signal
+        is given. If it times out, it returns false.
+  @param [in] queue The queue to pop from.
+  @param [in] item The item at the first poisition in the queue.
```
Should be `[out]`, right?

vpodzime commented on libutils/threaded_queue.h:
```diff
@@ -0,0 +1,132 @@
+  @param [in] queue The queue to pop from.
+  @param [in] item The item at the first poisition in the queue.
+  @param [in] timeout Timeout for blocking in seconds.
+  @return True on success, false if timed out or queue was empty.
```
*true*, it's a value not a random word here

vpodzime commented on libutils/threaded_queue.h:
```diff
@@ -0,0 +1,132 @@
+        signalled. If it times out, it returns 0 and sets *data_array to NULL.
+  @warning The pointer array will have to be freed manually.
+  @param [in] queue The queue to pop from.
+  @param [in] data_array Pointer to location to put popped elements.
```
Also `[out]`.
